google/auth/external_account.py

```diff
@@ -382,6 +397,50 @@ def with_token_uri(self, token_uri):
         kwargs.update(token_url=token_uri)
         return self.__class__(**kwargs)
 
+    def token_info_introspection(self, request):
+        """Introspection of token info.
+
+        If the credential object has a token_info_url provided, we can
+        introspect token info by requesting that endpoint.
+
+        Returns:
+            Mapping: The active token meta-information returned by the
+                introspection endpoint.
+
+        Raises:
+            google.auth.exceptions.InvalidResource:
+                If no token_info_url in config.
+            google.auth.exceptions.InvalidOperation:
+                If no token provided.
+            google.auth.exceptions.UserAccessTokenError:
+                If the credentials are invalid or expired.
+            google.auth.exceptions.TransportError:
+                If an error is encountered while calling the token
+                introspection endpoint.
+        """
+        # Token info introspection following RFC7662 standard:
+        #    https://datatracker.ietf.org/doc/html/rfc7662
+        #
+        if not self._token_info_url:
+            raise exceptions.InvalidResource("Missing token_info_url")
+
+        if not self.token:
+            raise exceptions.InvalidOperation("Missing token for introspection")
+
+        oauth_introspection_client = utils.TokenIntrospectionClient(
+            self._token_info_introspection, self._client_auth
+        )
+        if not request:
```

Comment:
I only see this pattern in one other place. Why fallback to requests instead of throwing an invalid argument exception?

Reply:
Actually as we have valid client auth (client id + client secret) and a sts issued token, we are OK to introspect the token. I think it doesn't matter that we have a request or not.
For here I just exposed one parameter in `token_info_introspection` method to keep consistency with other public methods.
I'd like to know your idea about this part.

Reply:
If I understood correctly, IMO I would not introduce the request parameter unless it is absolutely needed. 

You can always add a param without a breaking change, but they can never be removed.

Reply:
Now the case is:
In `oauth2.utils.TokenIntrospectionClient`, which will actually make the request there is no dependency on `requests` package. All the requests are passed in through the `introspection(request, token)` method.
If in `external_account.Credentials` we don't have dependency on the requests, then we must let it be provided by the caller.
Current implementation in `gcloud` is to having a `google.auth.request` session. And then that session's request calling the `IntrospectionClient` which implements in-house.